test/GPv2Transfer.test.ts

```diff
@@ -0,0 +1,160 @@
+    vault = await waffle.deployMockContract(deployer, IVault.abi);
+    token = await waffle.deployMockContract(deployer, IERC20.abi);
+  });
+
+  describe("transferToRecipient", () => {
+    const amount = ethers.utils.parseEther("13.37");
+
+    it("should transfer external amount to recipient", async () => {
+      await token.mock.transferFrom
+        .withArgs(traders[0].address, recipient.address, amount)
+        .returns(true);
+      await expect(
+        transfer.transferToRecipientTest(vault.address, recipient.address, [
+          {
+            account: traders[0].address,
+            token: token.address,
+            amount: amount,
+            useInternalBalance: false,
+          },
+        ]),
+      ).to.not.be.reverted;
+    });
+
+    it("should transfer internal amount to recipient", async () => {
+      await vault.mock.withdrawFromInternalBalance
+        .withArgs([
+          {
+            token: token.address,
+            amount,
+            sender: traders[0].address,
+            recipient: recipient.address,
+          },
+        ])
+        .returns();
+      await expect(
+        transfer.transferToRecipientTest(vault.address, recipient.address, [
+          {
+            account: traders[0].address,
+            token: token.address,
+            amount,
+            useInternalBalance: true,
+          },
+        ]),
+      ).to.not.be.reverted;
+    });
+
+    it("should transfer many external and internal amounts to recipient", async () => {
+      const transfers = traders.map((trader, i) => ({
+        account: trader.address,
+        token: token.address,
+        amount,
+        useInternalBalance: (i & 1) == 1,
+      }));
+
+      for (const { account } of transfers.filter(
+        (transfer) => !transfer.useInternalBalance,
+      )) {
+        await token.mock.transferFrom
+          .withArgs(account, recipient.address, amount)
+          .returns(true);
+      }
+      await vault.mock.withdrawFromInternalBalance
+        .withArgs(
+          transfers
+            .filter((transfer) => transfer.useInternalBalance)
+            .map(({ account }) => ({
+              token: token.address,
+              amount,
+              sender: account,
+              recipient: recipient.address,
+            })),
+        )
+        .returns();
+
+      await expect(
+        transfer.transferToRecipientTest(
+          vault.address,
+          recipient.address,
+          transfers,
+        ),
+      ).to.not.be.reverted;
+    });
+
+    it("reverts when mistakenly trying to transfer ETH", async () => {
+      await expect(
+        transfer.transferToRecipientTest(vault.address, recipient.address, [
+          {
+            account: traders[0].address,
+            token: BUY_ETH_ADDRESS,
+            amount: amount,
+            useInternalBalance: false,
+          },
+        ]),
+      ).to.be.revertedWith("GPv2: cannot transfer native ETH");
+    });
```

Comment:
I'd like to see this test include both states of `useInternalBalance`.

Reply:
That is the purpose of the `should transfer many external and internal amounts to recipient` test where we check that it works when both states of `userInternalBalance` are used in the same batched transfer.

Reply:
My goal was to enshrine that ETH should not be transferrable with the vault. In principle, with some small but reasonable changes the test you mention (and this) could succeed but the vault might still be used to transfer `0xee..ee`.

Reply:
Ahhhhhh, I misread this comment. Yes, for sure, we can check both internal and external balances for the `ETH` transfer reverts.